### PR TITLE
[Versions] Support legacy non-compliant PEP440 development versions 

### PR DESCRIPTION
**Pull request by enen92**
Fixes https://github.com/xbmc/addon-check/issues/224

Kodi supports non-PEP440 compliant version for development purposes, eg:

`1.0.1~beta`
`1.0.1~beta01`
etc

As those are not PEP440 compliant, the `packaging.version` module parses the version to `LegacyVersion` objects instead of `Version` objects. When the comparator compares both objects, `LegacyVersion` is always lower.

This is solved by ensuring that when a version string like those above is supplied to addon-checker (thus containing `~alpha` or `~beta`) and the resulting object is a `LegacyVersion`, the `AddonVersion` class does an implicit conversion to the PEP440 format: `x.x.xalphay`.

Tests added to mimic the ones in Kodi.

### CODE REVIEW

**Comment by enen92** [comment]
@SylvainCecchetto this should fix your issue. Thanks for reporting.

**Comment by SylvainCecchetto** [comment]
Tested on my side, it works fine! Thank you!

File: tests/test_versions.py [hunk 1]
```diff
@@ -18,11 +18,24 @@ def test_addonversion_simple_major():
...
+def test_kodi_non_compliant_pep440_builtin_versions():
+    # mimics the tests in TestAddonVersion.cpp (non PEP440 compliant)
+    # try to use PEP440 compliant versions whenever possible
+    assert AddonVersion("1.0.0-1") > AddonVersion("1.0.0")
+    assert AddonVersion("1.0.0~beta2") > AddonVersion("1.0.0~beta1")
+    assert AddonVersion("1.0.1~beta01") > AddonVersion("1.0.0")
+    assert AddonVersion("1.0.1~beta") > AddonVersion("1.0.0")
+    assert AddonVersion("1.0.0~beta") > AddonVersion("1.0.0~alpha")
+    assert AddonVersion("1.0.0~beta2") > AddonVersion("1.0.0~alpha")
+    assert AddonVersion("1.0.0~alpha3") > AddonVersion("1.0.0~alpha2")
```
Comment (Razzeee) [inline tests/test_versions.py:37]: Would tests with `+matrix.2` also make sense? Do we need to combine like `1.0.0+matrix.1~beta`?

File: kodi_addon_checker/versions.py [hunk 2]
```diff
@@ -6,13 +6,22 @@
...
+        self.version = parse(str(version).lower())
+        # non PEP440 compliant versions (legacy), for beta and alpha versions
+        # convert them into PEP440 format: 1.1.0~beta01 -> 1.1.0beta01
+        if self._islegacy_devversion:
+            self.version = Version(version.replace("~", ""))
+
+    @property
+    def _islegacy_devversion(self):
+        return isinstance(self.version, LegacyVersion) and \
+            ("~beta" or "~alpha" in self.version)
```
Comment (Razzeee) [inline kodi_addon_checker/versions.py:24]: I was under the impression that you could write whatever after `~` and I probably did at some point? At least for test builds, but I guess that was never supported by core?

File: tests/test_versions.py (hunk 1 above)
Comment (enen92) [reply tests/test_versions.py:37]: I don't think it makes much sense. `+matrix.1` is just a "wrapper" for the version available in matrix, so `1.0.0+matrix.1` is theoretically equivalent (at least for us) to version `1.0.0`. According to PEP440 is a local version identifier. 
Development versions in PEP440 are named `1.0.0betaX` or `1.0.0alphaX`.

So supposing that you have your `plugin.video.foo` in gotham (py3 compatible) with version 1.0.0:

version in gotham is `1.0.0`
version in matrix is `1.0.0+matrix.1`
version for development (external repo) is `1.0.1beta1`

`1.0.1beta1` > `1.0.0+matrix.1` > `1.0.0`

File: kodi_addon_checker/versions.py (hunk 2 above)
Comment (enen92) [reply kodi_addon_checker/versions.py:24]: Core tests only include `beta` and `alpha` so I'm honestly not sure what is supported or not. Note that this validates what we allow on the repo submissions  (and consider best practice) and not exactly what is theoretically possible. If some other edge case arrives, we can always improve on what we have now

File: tests/test_versions.py (hunk 1 above)
Comment (enen92) [reply tests/test_versions.py:37]: This is a good reminder I need to update the wiki when this versioning stuff is finished

**Comment by enen92** [comment]
(closed and reopen to trigger travis checks against our repo)
